src/Microsoft.DotNet.Interactive.CSharpProject/CSharpProjectKernel.cs

```diff
@@ -1,303 +1,318 @@
+                if (c == '\n')
+                {
+                    currentLine++;
+                    currentCharacter = 0;
+                }
+            }
+
+            return new LinePosition(currentLine, currentCharacter);
+        }
+
+        private async Task<Protocol.Workspace> GetWorkspaceWithCode(string code, int position = 0)
+        {
+            var updatedWorkspace = new Protocol.Workspace(
+                files: _workspace.Files,
+                buffers: _workspace.Buffers.Where(b => b.Id != _buffer.Id).Concat(new[] { new Protocol.Buffer(_buffer.Id, code, position: position) }).ToArray());
+            var inlinedWorkspace = await updatedWorkspace.InlineBuffersAsync();
+            return inlinedWorkspace;
+        }
+
+        private static IEnumerable<Diagnostic> GetDiagnostics(string code, Protocol.CompileResult result)
+        {
+            var diagnostics = Enumerable.Empty<Protocol.SerializableDiagnostic>();
+            var projectDiagnostics = Enumerable.Empty<Protocol.SerializableDiagnostic>();
+
+            if (result.Features.TryGetValue(nameof(Protocol.Diagnostics), out var candidateDiagnostics) &&
+                candidateDiagnostics is Protocol.Diagnostics diags)
+            {
+                diagnostics = diags;
+            }
+
+            if (result.Features.TryGetValue(nameof(Protocol.ProjectDiagnostics), out var candidateProjectDiagnostics) &&
+                candidateDiagnostics is Protocol.ProjectDiagnostics projectDiags)
+            {
+                projectDiagnostics = projectDiags;
+            }
+
+            var allDiagnostics = diagnostics.Concat(projectDiagnostics);
+
+            var finalDiagnostics = diagnostics.Select(d => new Diagnostic(new LinePositionSpan(GetLinePositionFromPosition(code, d.Start), GetLinePositionFromPosition(code, d.End)), ConvertSeverity(d.Severity), d.Id, d.Message));
+
+            return finalDiagnostics;
+        }
+
+        private static CodeAnalysis.DiagnosticSeverity ConvertSeverity(Protocol.DiagnosticSeverity severity)
+        {
+            switch (severity)
+            {
+                case Protocol.DiagnosticSeverity.Hidden:
+                    return CodeAnalysis.DiagnosticSeverity.Hidden;
+                case Protocol.DiagnosticSeverity.Info:
+                    return CodeAnalysis.DiagnosticSeverity.Info;
+                case Protocol.DiagnosticSeverity.Warning:
+                    return CodeAnalysis.DiagnosticSeverity.Warning;
+                case Protocol.DiagnosticSeverity.Error:
+                    return CodeAnalysis.DiagnosticSeverity.Error;
+                default:
+                    return CodeAnalysis.DiagnosticSeverity.Warning;
+            }
+        }
+
+        private void EnsureProjectOpened()
+        {
+            if (_workspaceServer is null || _workspace is null)
+            {
+                throw new Exception($"Project must be opened, send the command '{nameof(OpenProject)}' first.");

```

Comment:
We should never throw `Exception`.

The naming for these is a little unconventional. `Ensure*` usually means we're going to implicitly perform some prerequisite action. If you're going to throw, I would go with e.g. `ThrowIfProjectWasNotOpened`.